### PR TITLE
Nicer 404 error pages

### PR DESCRIPTION
**Pull request by daniel-beck**
This adds nicer 404 error pages to Jenkins. Preparation for https://issues.jenkins.io/browse/JENKINS-47400

<details>
<summary>Screenshots</summary>

### Before

Screenshot 2021-05-24 at 18 30 11

### After

Screenshot 2021-05-24 at 18 51 08

</details>

Since they are handled outside the usual filter chain (AFAIUI), even currently logged-in users would see a "login" link in the top right, so just remove that if we're currently serving an error page. This also applies to the existing catch-all error handler "Oops" page.

### Testing notes

If you get a Stapler error page on 404, run `org.kohsuke.stapler.Dispatcher.TRACE = false` in the script console.

### Proposed changelog entries

* Show nicer 404 error page when trying to access a URL that doesn't exist

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/resources/jenkins/model/Jenkins/404.jelly [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        <l:header />
+        <l:main-panel>
+            <h1 style="text-align: center">
+                <img src="${imagesURL}/rage.png" height="179" width="154"/> <span style="font-size:50px"><st:nbsp/>${%Oops!}</span>
+            </h1>
+            <div id="error-description" style="text-align: center">
+                <h2>${%title}</h2>
+                <p>
+                    <j:choose>
+                        <j:when test="${app.useSecurity}">
```
Comment (daniel-beck) [inline core/src/main/resources/jenkins/model/Jenkins/404.jelly:43]: This is a bit annoying -- we don't know who the current user is, so we cannot do permission checks to show a more tailored message.

File: core/src/main/resources/jenkins/model/Jenkins/404.jelly [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+<?jelly escape-by-default='true'?>
+<j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:t="/lib/hudson" xmlns:f="/lib/form" xmlns:i="jelly:fmt">
+    <!--
+      This is the page designated by web.xml to handle 404 errors. Show a nice error message and suggest logging in if applicable.
+      Generally kept very similar to 'oops.jelly'.
+    -->
+    <st:statusCode value="${response.getStatus()}" />
+    ${request.session.setAttribute('from', request.getAttribute('javax.servlet.error.request_uri'))}
+    <l:layout title="Jenkins" type="one-column">
```
Comment (timja) [inline core/src/main/resources/jenkins/model/Jenkins/404.jelly:33]: Should this say:
```suggestion
    <l:layout title="${%Page not found}" type="one-column">
```

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
re-triggered CI but the failing tests may be related:

`jenkins.security.ResourceDomainTest`

**Comment by daniel-beck** [comment]
> re-triggered CI but the failing tests may be related:
> 
> `jenkins.security.ResourceDomainTest`

I also think it is; the form validation expecting a 404 error page is probably getting tripped up by this.

**Comment by daniel-beck** [comment]
Well, this is awkward: The real problem for the test is that I don't expose `javax.servlet.error.message` in the 404 page, so the expected custom message for a properly set up resource root URL doesn't show. Easy to fix, but unfortunately looking at the 404 page on the resource root URL reveals another issue; since we do not serve anything other than static resources there, the shiny 404 is not actually shiny, as all styles and images fail to load. That's a bit more annoying.

**Comment by timja** [comment]
Is there a reason this is still in draft?

**Review by Wadeck — APPROVED** [review]
<3 

_(not tested manually but the change seems reasonable and will be nice)_

**Comment by daniel-beck** [comment]
> Is there a reason this is still in draft?

I actually wanted to accomplish something else here (foundation for removing the "Discover" permission), and while I was unable to do that here, I want to ensure this PR would not block future work on that.

File: core/src/main/resources/jenkins/model/Jenkins/404.properties [hunk 3]
```diff
@@ -0,0 +1,3 @@
+title = Error: {0}
```
Comment (timja) [inline core/src/main/resources/jenkins/model/Jenkins/404.properties:1]: 😢 the previous message you had in the screenshot looked nicer imo

image

**Pull request closed without merging**